File: examples/aws/python/00_auth_and_clients/client_with_profile.py
"""
Simple AWS Bedrock Converse example through the Lasso proxy using an AWS profile.

Required environment variables:
  LASSO_PROXY_ENDPOINT
  LASSO_X_API_KEY
  AWS_PROFILE
  AWS_REGION            # optional, defaults to us-east-1
  BEDROCK_MODEL_ID
"""

import json
import logging
import os

import boto3
import botocore.exceptions
from botocore.config import Config
from dotenv import load_dotenv


logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

load_dotenv()


# 1. Load configuration
lasso_proxy_endpoint = os.getenv("LASSO_PROXY_ENDPOINT")
lasso_api_key = os.getenv("LASSO_X_API_KEY")

profile_name = os.getenv("AWS_PROFILE")
region = os.getenv("AWS_REGION", "us-east-1")

model_id = os.getenv("BEDROCK_TEXT_MODEL_ID")


# 2. Create a boto3 session using an AWS profile
session = boto3.Session(
    profile_name=profile_name,
    region_name=region,
)


# 3. Create Bedrock client from the profile session and point it to Lasso
bedrock_client = session.client(
    "bedrock-runtime",
    endpoint_url=lasso_proxy_endpoint,
    config=Config(
        retries={"max_attempts": 1, "mode": "standard"},
        connect_timeout=5,
        read_timeout=12,
        user_agent_extra="lasso-proxy-example/1.0",
    ),
    verify=False
)


# 4. Add the Lasso API key header to every Bedrock request
def add_lasso_header(request, **kwargs):
    request.headers.add_header("lasso-x-api-key", lasso_api_key)


bedrock_client.meta.events.register_first(
    "before-sign.bedrock-runtime.*",
    add_lasso_header,
)


# 5. Call Bedrock Converse through Lasso
try:
    response = bedrock_client.converse(
        modelId=model_id,
        messages=[
            {
                "role": "user",
                "content": [
                    {
                        "text": "What is the ultimate answer to life, the universe, and everything?"
                    }
                ],
            }
        ],
        inferenceConfig={
            "maxTokens": 512,
        },
    )

    print(json.dumps(response, indent=2, default=str))

except botocore.exceptions.ClientError as error:
    logger.exception("Bedrock request failed: %s", error)

File: examples/aws/ruby/00_auth_and_clients/client_with_profile.rb
require 'aws-sdk-bedrockruntime'
require 'dotenv'
require 'json'

# -----------------------------------------------------------------------------
# Helpers
# -----------------------------------------------------------------------------

def require_env(name)
  value = ENV[name]

  raise "Missing required environment variable: #{name}" if value.nil? || value.strip.empty?

  value
end

def disable_ssl_verification?
  ENV['DISABLE_SSL_VERIFICATION'].to_s.casecmp('true').zero?
end

# -----------------------------------------------------------------------------
# Lasso AWS Bedrock Converse example using an AWS profile.
#
# This example sends Bedrock Runtime requests to the Lasso proxy. Lasso then
# forwards the request to Amazon Bedrock.
#
# Auth flow:
#   1. The AWS SDK loads credentials from the configured AWS profile.
#   2. The AWS SDK signs the Bedrock request with those profile credentials.
#   3. A small SDK plugin adds the Lasso API key header.
#   4. Lasso receives the signed Bedrock request and forwards it to AWS.
#
# Required .env values:
#   AWS_REGION
#   AWS_PROFILE
#   BEDROCK_TEXT_MODEL_ID
#   LASSO_PROXY_ENDPOINT
#   LASSO_X_API_KEY
#
# Optional .env value:
#   DISABLE_SSL_VERIFICATION=true
# -----------------------------------------------------------------------------

# -----------------------------------------------------------------------------
# 1. Load configuration
# -----------------------------------------------------------------------------

env_path = File.expand_path('../../../.env', __dir__)

raise "Missing .env file at #{env_path}" unless File.exist?(env_path)

Dotenv.load(env_path)

region = ENV.fetch('AWS_REGION', 'us-east-1')
profile_name = require_env('AWS_PROFILE')

model_id = require_env('BEDROCK_TEXT_MODEL_ID')
lasso_api_key = require_env('LASSO_X_API_KEY')
lasso_proxy_url = require_env('LASSO_PROXY_ENDPOINT').delete_suffix('/')

# The AWS SDK sends Bedrock Runtime requests to Lasso instead of AWS directly.
lasso_endpoint = "#{lasso_proxy_url}/v1/bedrock"

# -----------------------------------------------------------------------------
# 2. Add the Lasso API key to each Bedrock request
# -----------------------------------------------------------------------------

# The AWS SDK signs the request. This plugin only adds the Lasso API key header.
class LassoHeaderPlugin < Seahorse::Client::Plugin
  option(:lasso_api_key)

  class Handler < Seahorse::Client::Handler
    def call(context)
      context.http_request.headers['lasso-x-api-key'] =
        context.config.lasso_api_key

      @handler.call(context)
    end
  end

  handler(Handler, step: :build)
end

Aws::BedrockRuntime::Client.add_plugin(LassoHeaderPlugin)

# -----------------------------------------------------------------------------
# 3. Create the Bedrock Runtime client through Lasso
# -----------------------------------------------------------------------------

client = Aws::BedrockRuntime::Client.new(
  region: region,
  endpoint: lasso_endpoint,

  # The SDK loads credentials from ~/.aws/credentials and ~/.aws/config.
  profile: profile_name,

  lasso_api_key: lasso_api_key,

  # Local testing only.
  # Do not disable certificate validation in production.
  ssl_verify_peer: !disable_ssl_verification?
)

# -----------------------------------------------------------------------------
# 4. Send a Bedrock Converse request
# -----------------------------------------------------------------------------

response = client.converse(
  model_id: model_id,
  messages: [
    {
      role: 'user',
      content: [
        {
          text: 'In a single word, what is the answer to the great question of life, the universe and everything?'
        }
      ]
    }
  ],
  inference_config: {
    max_tokens: 512
  }
)

# -----------------------------------------------------------------------------
# 5. Print response
# -----------------------------------------------------------------------------

puts "HTTP status code: #{response.context.http_response.status_code}"

answer = response.to_h.dig(:output, :message, :content, 0, :text)

puts
puts 'Answer:'
puts answer
